### PR TITLE
RAB-1518: Make pause and resume working with remote storage

### PR DESCRIPTION
**Pull request by StevenVAIDIE**
**Description (for Contributor and Core Developer)**

In this PR,

We fixed the fact that some file are not exported to the remote storage when a job is paused and then resumed. This is because the UploadStep is not launched when a job is paused. As the written file are lost when we resume the job we cannot retrieve which file are not uploaded. 

We decided to: 
- Move the logic of archivist just after the step execution instead of been executed after the job is finished
- Now all file are in the archivist when we are in the upload step so now we can just copy file from archivist to the remote storage instead of regarding to the written file
 
**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php [hunk 1]
```diff
@@ -71,39 +68,29 @@ protected function doExecute(StepExecution $stepExecution)
...
-
-                $writtenFiles = array_merge($writtenFiles, $this->extractFileToTransferFromWriter($writer));
-            }
-        }
+        $filePath = strtr(
+            $this->fileWriterArchiver->getRelativeArchivePath($jobExecution),
+            [
+                '%filename%' => '',
+            ]
+        );
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php:76]: May be can move this logic into the FileWriterArchiver

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php [hunk 2]
```diff
@@ -71,39 +68,29 @@ protected function doExecute(StepExecution $stepExecution)
...
+        $dirname = str_replace(sys_get_temp_dir(), '', dirname($this->getPath($jobExecution)));
+
+        return array_map(function ($item) use ($dirname, $filePath) {
+            return new FileToTransfer(
+                $item,
+                'archivist',
+                $dirname . DIRECTORY_SEPARATOR . str_replace($filePath, '', $item),
+                false
+            );
+        }, iterator_to_array($this->fileWriterArchiver->getArchives($jobExecution, true)));
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php:87]:
```suggestion
        return array_map(static fn(string $filepath) => new FileToTransfer(
                $filePath,
                'archivist',
                $dirname . DIRECTORY_SEPARATOR . substr($filePath, 0, strlen($archiveFilePath)),
                false
            );
        }, iterator_to_array($this->fileWriterArchiver->getArchives($jobExecution, true)));
```

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php [hunk 3]
```diff
@@ -71,39 +68,29 @@ protected function doExecute(StepExecution $stepExecution)
...
-
-                $writer = $step->getWriter();
-                if (!$writer instanceof ArchivableWriterInterface) {
-                    continue;
-                }
-
-                $writtenFiles = array_merge($writtenFiles, $this->extractFileToTransferFromWriter($writer));
-            }
-        }
+        $filePath = strtr(
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php:71]:
```suggestion
        $archiveFilePath = strtr(
```

File: src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php [hunk 4]
```diff
@@ -71,39 +68,29 @@ protected function doExecute(StepExecution $stepExecution)
...
+            $this->fileWriterArchiver->getRelativeArchivePath($jobExecution),
+            [
+                '%filename%' => '',
+            ]
+        );
 
-        return $writtenFiles;
+        $dirname = str_replace(sys_get_temp_dir(), '', dirname($this->getPath($jobExecution)));
```
Comment (StevenVAIDIE) [inline src/Akeneo/Platform/Bundle/ImportExportBundle/Infrastructure/Step/UploadStep.php:78]: In $this->getPath() we add sys_get_temp_dir() and here we remove it we can just remove it here and in $this->getPath()

File: src/Akeneo/Tool/Component/Batch/Job/Job.php [hunk 5]
```diff
@@ -151,6 +151,7 @@ final public function execute(JobExecution $jobExecution): void
                 $this->jobRepository->updateJobExecution($jobExecution);
             }
 
+
```
Comment (StevenVAIDIE) [inline src/Akeneo/Tool/Component/Batch/Job/Job.php:154]: File to reset

File: src/Akeneo/Tool/Component/Batch/Step/AbstractStep.php [hunk 6]
```diff
@@ -95,6 +95,7 @@ final public function execute(StepExecution $stepExecution)
             } else {
                 // Need to upgrade here not set, in case the execution was stopped
                 $stepExecution->upgradeStatus(BatchStatus::COMPLETED);
+
                 $this->dispatchStepExecutionEvent(EventInterface::STEP_EXECUTION_SUCCEEDED, $stepExecution);
```
Comment (StevenVAIDIE) [inline src/Akeneo/Tool/Component/Batch/Step/AbstractStep.php:99]: File to reset

File: src/Akeneo/Tool/Component/Batch/Step/ItemStep.php [hunk 7]
```diff
@@ -92,7 +92,9 @@ public function doExecute(StepExecution $stepExecution)
             $stepExecution->setTotalItems($this->getCountFromTrackableItemReader());
         }
 
+//        $this->batchSize = 1;
```
Comment (StevenVAIDIE) [inline src/Akeneo/Tool/Component/Batch/Step/ItemStep.php:95]: File to reset

File: src/Akeneo/Tool/Component/Connector/Archiver/AbstractFilesystemArchiver.php [hunk 8]
```diff
@@ -71,4 +71,14 @@ protected function getRelativeArchivePath(JobExecution $jobExecution): string
...
     }
+
+    private function supportJobExecution(JobExecution $jobExecution): bool
+    {
+        foreach ($jobExecution->getStepExecutions() as $stepExecution) {
+            if ($this->supports($stepExecution)) {
+                return true;
+            }
+        }
+        return false;
```
Comment (StevenVAIDIE) [inline src/Akeneo/Tool/Component/Connector/Archiver/AbstractFilesystemArchiver.php:82]: Coding style
```suggestion
        foreach ($jobExecution->getStepExecutions() as $stepExecution) {
            if ($this->supports($stepExecution)) {
                return true;
            }
        }

        return false;
```

File: src/Akeneo/Tool/Component/Connector/Archiver/AbstractFilesystemArchiver.php [hunk 9]
```diff
@@ -22,7 +22,7 @@ abstract class AbstractFilesystemArchiver implements ArchiverInterface
      */
     public function getArchives(JobExecution $jobExecution, bool $deep = false): iterable
     {
-        if (!$this->supports($jobExecution)) {
+        if (!$this->supportJobExecution($jobExecution)) {
```
Comment (ValentinMumble) [inline src/Akeneo/Tool/Component/Connector/Archiver/AbstractFilesystemArchiver.php:25]:
```suggestion
        if (!$this->supportsJobExecution($jobExecution)) {
```

**Review by ValentinMumble — APPROVED** [review]

**Review by marinfazioakeneo — APPROVED** [review]

**Pull request merged**